OrbitBase/include/OrbitBase/ThreadPool.h

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_THREAD_POOL_H_
+#define ORBIT_BASE_THREAD_POOL_H_
+
+#include <memory>
+
+#include "OrbitBase/Action.h"
+
+// This class implements a Thread Pool. Thread Pool allows to execute
+// actions in another thread without need to manage creation and destruction
+// of that thread.
+//
+// Usage example:
+//
+// /* Schedule an action on the thread-pool */
+// thread_poool->Schedule(CreateAction([data]{
+//   DoSomethingWith(data);
+// }));
+//
+// /* Shutdown thread-pool, blocks until all currently executed actions are
+//    complete and stops all worker threads */
+//
+// thread_pool->Shutdown();
+// thread_pool->Wait();
+//
+class ThreadPool {
+ public:
+  ThreadPool() = default;
+  virtual ~ThreadPool() = default;
+
+  // Schedules the action to be performed on the main thread.
+  virtual void Schedule(std::unique_ptr<Action> action) = 0;
+
+  template <typename F>
+  void Schedule(F&& functor) {
+    Schedule(CreateAction(std::forward<F>(functor)));
+  }
+
+  // Initiates shutdown, any Schedule after this call will fail.
+  virtual void Shutdown() = 0;
+
+  // Wait until all tasks are complete. Call this after Shutdown
```

Comment:
I would clarify the documentation a bit here: 
- Is it ok to call this without having called Shutdown() first? (I'd assume, yes, but not clear from the comment.)
- Do I have to call this after Shutdown()?

Reply:
Clarified - added ShutdownAndWait() method for convenience.